File: packages/nerva-py/nerva/testkit/assertions.py
"""Assertion helpers for Nerva testkit.

Concise, readable assertions for the most common test scenarios.
Each function inspects spy call records and raises ``AssertionError``
with a descriptive message on failure.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from nerva.testkit.builders import TestOrchestratorResult
    from nerva.testkit.spies import (
        SpyMemory,
        SpyPolicy,
        SpyRouter,
        SpyRuntime,
        SpyToolManager,
    )


def assert_routed_to(spy_router: SpyRouter, handler_name: str) -> None:
    """Assert that the router's most recent classify() selected the given handler.

    Args:
        spy_router: The SpyRouter to inspect.
        handler_name: Expected handler name.

    Raises:
        AssertionError: If no classify calls recorded or handler doesn't match.
    """
    assert len(spy_router.classify_calls) > 0, "SpyRouter has no recorded classify calls"
    last_call = spy_router.classify_calls[-1]
    best = last_call.result.best_handler
    actual_name = best.name if best else None
    assert actual_name == handler_name, (
        f"Expected route to '{handler_name}', got '{actual_name}'"
    )


def assert_handler_invoked(
    spy_runtime: SpyRuntime,
    handler_name: str,
    *,
    message: str | None = None,
) -> None:
    """Assert that the runtime invoked a specific handler.

    Args:
        spy_runtime: The SpyRuntime to inspect.
        handler_name: Expected handler name.
        message: If provided, also assert the input message matches.

    Raises:
        AssertionError: If no matching invoke call is found.
    """
    matching = [c for c in spy_runtime.invoke_calls if c.handler == handler_name]
    assert len(matching) > 0, (
        f"Handler '{handler_name}' was never invoked. "
        f"Invoked handlers: {[c.handler for c in spy_runtime.invoke_calls]}"
    )
    if message is not None:
        messages = [c.input.message for c in matching]
        assert message in messages, (
            f"Handler '{handler_name}' was invoked but not with message '{message}'. "
            f"Messages seen: {messages}"
        )


def assert_policy_allowed(spy_policy: SpyPolicy) -> None:
    """Assert that the most recent policy evaluation allowed the action.

    Args:
        spy_policy: The SpyPolicy to inspect.

    Raises:
        AssertionError: If no evaluate calls or the last one denied.
    """
    assert len(spy_policy.evaluate_calls) > 0, "SpyPolicy has no recorded evaluate calls"
    last_call = spy_policy.evaluate_calls[-1]
    assert last_call.result.allowed, (
        f"Expected policy to allow, but it denied with reason: {last_call.result.reason}"
    )


def assert_policy_denied(spy_policy: SpyPolicy, *, reason: str | None = None) -> None:
    """Assert that the most recent policy evaluation denied the action.

    Args:
        spy_policy: The SpyPolicy to inspect.
        reason: If provided, also assert the denial reason matches.

    Raises:
        AssertionError: If no evaluate calls or the last one allowed.
    """
    assert len(spy_policy.evaluate_calls) > 0, "SpyPolicy has no recorded evaluate calls"
    last_call = spy_policy.evaluate_calls[-1]
    assert not last_call.result.allowed, "Expected policy to deny, but it allowed"
    if reason is not None:
        assert last_call.result.reason == reason, (
            f"Expected denial reason '{reason}', got '{last_call.result.reason}'"
        )


def assert_memory_stored(spy_memory: SpyMemory, *, content: str | None = None) -> None:
    """Assert that at least one memory store() call was made.

    Args:
        spy_memory: The SpyMemory to inspect.
        content: If provided, assert that at least one stored event has this content.

    Raises:
        AssertionError: If no store calls or no matching content.
    """
    assert len(spy_memory.store_calls) > 0, "SpyMemory has no recorded store calls"
    if content is not None:
        contents = [c.event.content for c in spy_memory.store_calls]
        assert content in contents, (
            f"No stored event with content '{content}'. Stored: {contents}"
        )


def assert_memory_recalled(spy_memory: SpyMemory, *, query: str | None = None) -> None:
    """Assert that at least one memory recall() call was made.

    Args:
        spy_memory: The SpyMemory to inspect.
        query: If provided, assert that at least one recall used this query.

    Raises:
        AssertionError: If no recall calls or no matching query.
    """
    assert len(spy_memory.recall_calls) > 0, "SpyMemory has no recorded recall calls"
    if query is not None:
        queries = [c.query for c in spy_memory.recall_calls]
        assert query in queries, (
            f"No recall with query '{query}'. Queries: {queries}"
        )


def assert_tool_called(
    spy_tools: SpyToolManager,
    tool_name: str,
    *,
    args: dict[str, object] | None = None,
) -> None:
    """Assert that a specific tool was called.

    Args:
        spy_tools: The SpyToolManager to inspect.
        tool_name: Expected tool name.
        args: If provided, assert the tool was called with these args.

    Raises:
        AssertionError: If no matching tool call is found.
    """
    matching = [c for c in spy_tools.call_calls if c.tool_name == tool_name]
    assert len(matching) > 0, (
        f"Tool '{tool_name}' was never called. "
        f"Called tools: {[c.tool_name for c in spy_tools.call_calls]}"
    )
    if args is not None:
        actual_args = [c.args for c in matching]
        assert args in actual_args, (
            f"Tool '{tool_name}' was called but not with args {args}. "
            f"Args seen: {actual_args}"
        )


def assert_no_unconsumed_expectations(result: TestOrchestratorResult) -> None:
    """Assert that all spies have consumed their expectations.

    Args:
        result: The TestOrchestratorResult to inspect.

    Raises:
        AssertionError: If any spy has pending expectations.
    """
    result.verify_all_expectations_consumed()


def assert_pipeline_order(
    result: TestOrchestratorResult,
    expected_order: list[str],
) -> None:
    """Assert that primitives were called in the expected order.

    Compares timestamps of the earliest call to each primitive.
    Valid primitive names: ``"router"``, ``"runtime"``, ``"responder"``,
    ``"memory"``, ``"policy"``, ``"tools"``.

    Args:
        result: The TestOrchestratorResult to inspect.
        expected_order: List of primitive names in expected execution order.

    Raises:
        AssertionError: If the actual order doesn't match.
    """
    timestamp_map: dict[str, float] = {}

    if result.router.classify_calls:
        timestamp_map["router"] = result.router.classify_calls[0].timestamp
    if result.runtime.invoke_calls:
        timestamp_map["runtime"] = result.runtime.invoke_calls[0].timestamp
    if result.responder.format_calls:
        timestamp_map["responder"] = result.responder.format_calls[0].timestamp
    if result.memory.recall_calls:
        timestamp_map["memory"] = result.memory.recall_calls[0].timestamp
    if result.policy.evaluate_calls:
        timestamp_map["policy"] = result.policy.evaluate_calls[0].timestamp
    if result.tools.discover_calls or result.tools.call_calls:
        tool_timestamps = (
            [c.timestamp for c in result.tools.discover_calls]
            + [c.timestamp for c in result.tools.call_calls]
        )
        timestamp_map["tools"] = min(tool_timestamps)

    actual_order = [
        name for name in expected_order if name in timestamp_map
    ]
    actual_sorted = sorted(actual_order, key=lambda n: timestamp_map[n])

    assert actual_sorted == actual_order, (
        f"Expected pipeline order {actual_order}, but got {actual_sorted}"
    )

File: packages/nerva-js/src/testkit/assertions.ts
/**
 * Assertion helpers for Nerva testkit.
 *
 * Concise, readable assertions for the most common test scenarios.
 * Each function inspects spy call records and throws `Error` with a
 * descriptive message on failure.
 *
 * @module testkit/assertions
 */

import type { SpyRouter } from "./spies.js";
import type { SpyRuntime } from "./spies.js";
import type { SpyPolicy } from "./spies.js";
import type { SpyMemory } from "./spies.js";
import type { SpyToolManager } from "./spies.js";
import type { TestOrchestratorResult } from "./builders.js";

/**
 * Assert that the router's most recent classify() selected the given handler.
 *
 * @param spyRouter - The SpyRouter to inspect.
 * @param handlerName - Expected handler name.
 * @throws {Error} If no classify calls recorded or handler doesn't match.
 */
export function assertRoutedTo(
  spyRouter: SpyRouter,
  handlerName: string,
): void {
  if (spyRouter.classifyCalls.length === 0) {
    throw new Error("SpyRouter has no recorded classify calls");
  }
  const lastCall = spyRouter.classifyCalls[spyRouter.classifyCalls.length - 1]!;
  const actualName = lastCall.result.bestHandler?.name ?? null;
  if (actualName !== handlerName) {
    throw new Error(
      `Expected route to '${handlerName}', got '${actualName}'`,
    );
  }
}

/**
 * Assert that the runtime invoked a specific handler.
 *
 * @param spyRuntime - The SpyRuntime to inspect.
 * @param handlerName - Expected handler name.
 * @param options - Optional message filter.
 * @throws {Error} If no matching invoke call is found.
 */
export function assertHandlerInvoked(
  spyRuntime: SpyRuntime,
  handlerName: string,
  options?: { message?: string },
): void {
  const matching = spyRuntime.invokeCalls.filter(
    (c) => c.handler === handlerName,
  );
  if (matching.length === 0) {
    const invoked = spyRuntime.invokeCalls.map((c) => c.handler);
    throw new Error(
      `Handler '${handlerName}' was never invoked. Invoked: ${JSON.stringify(invoked)}`,
    );
  }
  if (options?.message !== undefined) {
    const messages = matching.map((c) => c.input.message);
    if (!messages.includes(options.message)) {
      throw new Error(
        `Handler '${handlerName}' was invoked but not with message '${options.message}'. Messages: ${JSON.stringify(messages)}`,
      );
    }
  }
}

/**
 * Assert that the most recent policy evaluation allowed the action.
 *
 * @param spyPolicy - The SpyPolicy to inspect.
 * @throws {Error} If no evaluate calls or the last one denied.
 */
export function assertPolicyAllowed(spyPolicy: SpyPolicy): void {
  if (spyPolicy.evaluateCalls.length === 0) {
    throw new Error("SpyPolicy has no recorded evaluate calls");
  }
  const lastCall = spyPolicy.evaluateCalls[spyPolicy.evaluateCalls.length - 1]!;
  if (!lastCall.result.allowed) {
    throw new Error(
      `Expected policy to allow, but it denied with reason: ${lastCall.result.reason}`,
    );
  }
}

/**
 * Assert that the most recent policy evaluation denied the action.
 *
 * @param spyPolicy - The SpyPolicy to inspect.
 * @param options - Optional reason filter.
 * @throws {Error} If no evaluate calls or the last one allowed.
 */
export function assertPolicyDenied(
  spyPolicy: SpyPolicy,
  options?: { reason?: string },
): void {
  if (spyPolicy.evaluateCalls.length === 0) {
    throw new Error("SpyPolicy has no recorded evaluate calls");
  }
  const lastCall = spyPolicy.evaluateCalls[spyPolicy.evaluateCalls.length - 1]!;
  if (lastCall.result.allowed) {
    throw new Error("Expected policy to deny, but it allowed");
  }
  if (options?.reason !== undefined && lastCall.result.reason !== options.reason) {
    throw new Error(
      `Expected denial reason '${options.reason}', got '${lastCall.result.reason}'`,
    );
  }
}

/**
 * Assert that at least one memory store() call was made.
 *
 * @param spyMemory - The SpyMemory to inspect.
 * @param options - Optional content filter.
 * @throws {Error} If no store calls or no matching content.
 */
export function assertMemoryStored(
  spyMemory: SpyMemory,
  options?: { content?: string },
): void {
  if (spyMemory.storeCalls.length === 0) {
    throw new Error("SpyMemory has no recorded store calls");
  }
  if (options?.content !== undefined) {
    const contents = spyMemory.storeCalls.map((c) => c.event.content);
    if (!contents.includes(options.content)) {
      throw new Error(
        `No stored event with content '${options.content}'. Stored: ${JSON.stringify(contents)}`,
      );
    }
  }
}

/**
 * Assert that at least one memory recall() call was made.
 *
 * @param spyMemory - The SpyMemory to inspect.
 * @param options - Optional query filter.
 * @throws {Error} If no recall calls or no matching query.
 */
export function assertMemoryRecalled(
  spyMemory: SpyMemory,
  options?: { query?: string },
): void {
  if (spyMemory.recallCalls.length === 0) {
    throw new Error("SpyMemory has no recorded recall calls");
  }
  if (options?.query !== undefined) {
    const queries = spyMemory.recallCalls.map((c) => c.query);
    if (!queries.includes(options.query)) {
      throw new Error(
        `No recall with query '${options.query}'. Queries: ${JSON.stringify(queries)}`,
      );
    }
  }
}

/**
 * Assert that a specific tool was called.
 *
 * @param spyTools - The SpyToolManager to inspect.
 * @param toolName - Expected tool name.
 * @param options - Optional args filter.
 * @throws {Error} If no matching tool call is found.
 */
export function assertToolCalled(
  spyTools: SpyToolManager,
  toolName: string,
  options?: { args?: Record<string, unknown> },
): void {
  const matching = spyTools.callCalls.filter((c) => c.toolName === toolName);
  if (matching.length === 0) {
    const called = spyTools.callCalls.map((c) => c.toolName);
    throw new Error(
      `Tool '${toolName}' was never called. Called: ${JSON.stringify(called)}`,
    );
  }
  if (options?.args !== undefined) {
    const argsMatch = matching.some(
      (c) => JSON.stringify(c.args) === JSON.stringify(options.args),
    );
    if (!argsMatch) {
      const actualArgs = matching.map((c) => c.args);
      throw new Error(
        `Tool '${toolName}' was called but not with args ${JSON.stringify(options.args)}. Args: ${JSON.stringify(actualArgs)}`,
      );
    }
  }
}

/**
 * Assert that all spies have consumed their expectations.
 *
 * @param result - The TestOrchestratorResult to inspect.
 * @throws {Error} If any spy has pending expectations.
 */
export function assertNoUnconsumedExpectations(
  result: TestOrchestratorResult,
): void {
  result.verifyAllExpectationsConsumed();
}

/**
 * Assert that primitives were called in the expected order.
 *
 * @param result - The TestOrchestratorResult to inspect.
 * @param expectedOrder - List of primitive names in expected execution order.
 * @throws {Error} If the actual order doesn't match.
 */
export function assertPipelineOrder(
  result: TestOrchestratorResult,
  expectedOrder: string[],
): void {
  const timestampMap: Record<string, number> = {};

  if (result.router.classifyCalls.length > 0) {
    timestampMap["router"] = result.router.classifyCalls[0]!.timestamp;
  }
  if (result.runtime.invokeCalls.length > 0) {
    timestampMap["runtime"] = result.runtime.invokeCalls[0]!.timestamp;
  }
  if (result.responder.formatCalls.length > 0) {
    timestampMap["responder"] = result.responder.formatCalls[0]!.timestamp;
  }
  if (result.memory.recallCalls.length > 0) {
    timestampMap["memory"] = result.memory.recallCalls[0]!.timestamp;
  }
  if (result.policy.evaluateCalls.length > 0) {
    timestampMap["policy"] = result.policy.evaluateCalls[0]!.timestamp;
  }
  const toolTimestamps = [
    ...result.tools.discoverCalls.map((c) => c.timestamp),
    ...result.tools.callCalls.map((c) => c.timestamp),
  ];
  if (toolTimestamps.length > 0) {
    timestampMap["tools"] = Math.min(...toolTimestamps);
  }

  const actualOrder = expectedOrder.filter((name) => name in timestampMap);
  const actualSorted = [...actualOrder].sort(
    (a, b) => (timestampMap[a] ?? 0) - (timestampMap[b] ?? 0),
  );

  if (JSON.stringify(actualSorted) !== JSON.stringify(actualOrder)) {
    throw new Error(
      `Expected pipeline order ${JSON.stringify(actualOrder)}, but got ${JSON.stringify(actualSorted)}`,
    );
  }
}
